File: Baekjun/7.2/오아시스 재결합/오아시스재결합.py
# 둘 사이에 둘 중에 하나보다 큰 녀석이 있으면 안 된다.
N = int(input())
# (height, cnt)
stack = []
ans = 0
nums = [int(input()) for _ in range(N)]

for i in range(N):
    while stack and nums[i] > stack[-1][0]:
        ans += stack.pop()[1]

    if not stack:
        stack.append((nums[i], 1))
        continue

    # 현재 스택의 끝값과 같을 때
    if stack[-1][0] == nums[i]:
        cnt = stack.pop()[1]
        ans += cnt

        if stack:
            ans += 1

        stack.append((nums[i], cnt+1))
    else:
        stack.append((nums[i], 1))
        ans += 1

print(ans)

File: Baekjun/7.2/오아시스 재결합/오아시스재결합.swift
var stack: [(Int, Int)] = []
var nums: [Int] = []
var n = Int(readLine()!)
var ans = 0

for _ in 0..<n! {
    if let val = readLine() {
        nums.append(Int(val)!)
    }
}

for i in 0..<n! {
    while !stack.isEmpty && nums[i] > stack[stack.count-1].0 {
        ans += stack.popLast()!.1
    }
    if stack.isEmpty {
        stack.append((nums[i], 1))
        continue
    }
    
    if stack[stack.count-1].0 == nums[i] {
        let cnt = stack.popLast()!.1
        ans += cnt
        
        if !stack.isEmpty {
            ans += 1
        }
        
        stack.append((nums[i], cnt + 1))
    } else {
        ans += 1
        stack.append((nums[i], 1))
    }
}

print(ans)
